File: problems/3122_minimum_number_of_operations_to_satisfy_conditions/Solution.java
import java.util.*;

class Solution {
    public int minimumOperations(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] freq = new int[n][10];
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                freq[j][grid[i][j]]++;
        int[][] dp = new int[n][10];
        for (int[] row : dp) Arrays.fill(row, Integer.MAX_VALUE);
        for (int d = 0; d < 10; d++) dp[0][d] = m - freq[0][d];
        for (int j = 1; j < n; j++)
            for (int d = 0; d < 10; d++) {
                int cost = m - freq[j][d];
                for (int pd = 0; pd < 10; pd++)
                    if (pd != d) dp[j][d] = Math.min(dp[j][d], dp[j-1][pd] + cost);
            }
        int res = Integer.MAX_VALUE;
        for (int d = 0; d < 10; d++) res = Math.min(res, dp[n-1][d]);
        return res;
    }
}

File: problems/3122_minimum_number_of_operations_to_satisfy_conditions/solution.py
from typing import List


class Solution:
    def minimumOperations(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        freq = [[0] * 10 for _ in range(n)]
        for i in range(m):
            for j in range(n):
                freq[j][grid[i][j]] += 1
        dp = [[float('inf')] * 10 for _ in range(n)]
        for d in range(10):
            dp[0][d] = m - freq[0][d]
        for j in range(1, n):
            for d in range(10):
                cost = m - freq[j][d]
                for pd in range(10):
                    if pd != d:
                        dp[j][d] = min(dp[j][d], dp[j-1][pd] + cost)
        return min(dp[n-1])
